contract/contracts/expense-account/ExpenseAccountEIP712.sol

```diff
@@ -52,61 +126,123 @@ contract ExpenseAccountEIP712 is
         __EIP712_init("CNCExpenseAccount", "1");
     }
 
+    /**
+     * @dev Computes the hash of a single BudgetData item.
+     * @param budgetData The BudgetData struct to hash.
+     * @return The hash of the BudgetData item.
+     */
+    function singleBudgetDataHash (BudgetData calldata budgetData) private pure returns( bytes32 ) {
+        return keccak256(abi.encode(
+            BUDGETDATA_TYPEHASH,
+            budgetData.budgetType,
+            budgetData.value
+        ));
+    }
+
+    /**
+     * @dev Computes the hash of an array of BudgetData items.
+     * @param budgetData The BudgetData items to hash.
+     * @return The hash of the BudgetData items.
+     */
+    function budgetDataHash (BudgetData[] calldata budgetData) private pure returns( bytes32 ) {
+        bytes32[] memory hashes = new bytes32[](budgetData.length);
+        for (uint8 i = 0; i < budgetData.length; i++) {
+            hashes[i] = singleBudgetDataHash(budgetData[i]);
+        }
+        return keccak256(abi.encodePacked(hashes)); // Hash the array of hashes
+    }
+
+    /**
+     * @dev Computes the hash of an BudgetLimit struct which is the object that
+     * was signed by the contract owner.
+     * @param limit The BudgetLimit struct to hash.
+     * @return The hash of the BudgetLimit struct.
+     */
     function budgetLimitHash (BudgetLimit calldata limit) private pure returns( bytes32 ) {
         return keccak256(abi.encode(
             BUDGETLIMIT_TYPEHASH,
             limit.approvedAddress,
-            limit.budgetType,
-            limit.value,
+            budgetDataHash(limit.budgetData),
             limit.expiry
         ));
     }
 
+    /**
+     * @notice Allows an employee to withdraw their wages.
+     * @dev Hourly rate is in ether so it needs to be multiplied
+     * by 10 ** 18
+     * @param to The address to transfer to.
+     * @param amount The amount to transfer.
+     * @param limit The BudgetLimit struct that was signed by the contract owner
+     * @param signature The ECDSA signature.
+     *
+     * Requirements:
+     * - The caller must be the member specified in the budget limit.
+     * - The budget limit must be signed by the contract owner.
+     * - The number of transactions must not exceed the specified amount.
+     * - The total amount withdrawn must not exceed the allowed amount per period.
+     * - The amount being transferred must not exceed the allowed amount per transaction.
+     * - The budgetData must not be an empty array.
+     * - The budgetData must be a valid budget type.
+     * - The contract must not be paused.
+     *
+     * Emits a {Withdraw} event.
+     */
     function transfer(
```

Comment:
i think you should refactor this function since it's getting more longer. please review my code below

```solidity
 function transfer(
        address to,
        uint256 amount, 
        BudgetLimit calldata limit, 
        bytes calldata signature
    ) external whenNotPaused nonReentrant {
        require(msg.sender == limit.approvedAddress, "Withdrawer not approved");
        require(to != address(0), "Address required");
        require(amount > 0, "Amount must be greater than zero");
        require(limit.budgetData.length > 0, "Empty budget data");

        bytes32 digest = _hashTypedDataV4(budgetLimitHash(limit));
        address signer = digest.recover(signature);

        if (signer != owner()) {
            revert UnauthorizedAccess(owner(), signer);
        }

        require(block.timestamp <= limit.expiry, "Authorization expired");

        bytes32 sigHash = keccak256(signature);
        Balance storage balance = balances[sigHash];

        _checkAndUpdateBudgetData(limit.budgetData, balance, amount);

        payable(to).sendValue(amount);
        emit Transfer(limit.approvedAddress, to, amount);
    }

    function _checkAndUpdateBudgetData(BudgetData[] calldata budgetData, Balance storage balance, uint256 amount) private {
        for (uint8 i = 0; i < budgetData.length; i++) {
            BudgetData memory budget = budgetData[i];
            if (budget.budgetType == BudgetType.TransactionsPerPeriod) {
                require(balance.transactionCount < budget.value, "Transaction limit reached");
                balance.transactionCount++;
            } else if (budget.budgetType == BudgetType.AmountPerPeriod) {
                require(balance.amountWithdrawn + amount <= budget.value, "Amount per period exceeded");
                balance.amountWithdrawn += amount;
            } else if (budget.budgetType == BudgetType.AmountPerTransaction) {
                require(amount <= budget.value, "Amount per transaction exceeded");
            }
        }
    }
```